Unmarshalizing Fails After Line Comments

# Relevant Version
- the latest [HEAD](https://github.com/zoncoen/scenarigo/commit/5f0105b84f496f5fc66754ff64e58061aea58d39) on main branch.
    - No issue on the tagged latest `v0.14.2`.

# Symptom
The scenario with comments like below fails to unmarshal the yaml file.

Given the scenario like:
```yml
title: get scenarigo repository
steps:
- title: expect 404
  protocol: http
  request:
    method: GET
    url: "https://api.github.com/repos/zoncoen/aaaaaaaaaaaaa"
    body:
      # message: this comment does evil
  expect:
    code: Not Found
```


Here is the error,
```
/Users/username/code/repo/scenarigo/.bin/scenarigo run
--- FAIL: scenario/github.yaml (0.40s)
    --- FAIL: scenario/github.yaml/get_scenarigo_repository (0.40s)
        --- FAIL: scenario/github.yaml/get_scenarigo_repository/expect_404 (0.40s)
                request:
                  method: GET
                  url: https://api.github.com/repos/zoncoen/aaaaaaaaaaaaa
                  header:
                    User-Agent:
                    - scenarigo/v0.14.2-dev
                response:
                  header:
                    Access-Control-Allow-Origin:
                    - "*"
                    Access-Control-Expose-Headers:
                    - ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset
                    Content-Encoding:
                    - gzip
                    Content-Length:
                    - "112"
                    Content-Security-Policy:
                    - default-src 'none'
                    Content-Type:
                    - application/json; charset=utf-8
                    Date:
                    - Thu, 01 Jun 2023 05:41:32 GMT
                    Referrer-Policy:
                    - origin-when-cross-origin, strict-origin-when-cross-origin
                    Server:
                    - GitHub.com
                    Strict-Transport-Security:
                    - max-age=31536000; includeSubdomains; preload
                    Vary:
                    - Accept-Encoding, Accept, X-Requested-With
                    X-Content-Type-Options:
                    - nosniff
                    X-Frame-Options:
                    - deny
                    X-Github-Api-Version-Selected:
                    - 2022-11-28
                    X-Github-Media-Type:
                    - github.v3; format=json
                    X-Github-Request-Id:
                    - 0505:0BE5:DBA2D8:E77D3C:64782F8C
                    X-Ratelimit-Limit:
                    - "60"
                    X-Ratelimit-Remaining:
                    - "48"
                    X-Ratelimit-Reset:
                    - "1685599002"
                    X-Ratelimit-Resource:
                    - core
                    X-Ratelimit-Used:
                    - "12"
                    X-Xss-Protection:
                    - "0"
                  body:
                    documentation_url: "https://docs.github.com/rest/reference/repos#get-a-repository"
                    message: Not Found
                elapsed time: 0.401880 sec
                .steps[0].expect.code: expected 200 but got Not Found
FAIL
FAIL	scenario/github.yaml	0.403s
FAIL
```

Scenario fails with `expected 200`, where `200` is a [default code](https://github.com/zoncoen/scenarigo/blob/5f0105b84f496f5fc66754ff64e58061aea58d39/protocol/http/expect.go#L22), indicating that the lines after the comment (`# message: this comment does evil`) are dropped somehow.

--- a/runner_test.go
+++ b/runner_test.go
@@ -493,25 +493,6 @@ func TestRunner_Dump(t *testing.T) {
 					},
 				},
 			},
-			"disable ytt integration": {
-				config: &schema.Config{
-					Scenarios: []string{"testdata/ytt.yaml"},
-				},
-				expect: `schemaVersion: scenario/v1
-title: echo
-vars:
-  message: null
-steps:
-- title: POST /say
-  protocol: http
-  request:
-    body:
-      message: "{{vars.message}}"
-  expect:
-    body:
-      message: "{{request.message}}"
-`,
-			},
 			"enable ytt integration": {
 				config: &schema.Config{
 					Scenarios: []string{"testdata/ytt.yaml"},
@@ -527,25 +508,6 @@ steps:
 title: echo
 vars:
   message: hello
-steps:
-- title: POST /say
-  protocol: http
-  request:
-    body:
-      message: "{{vars.message}}"
-  expect:
-    body:
-      message: "{{request.message}}"
-`,
-			},
-			"invalid but disable ytt integration": {
-				config: &schema.Config{
-					Scenarios: []string{"testdata/ytt_invalid.yaml"},
-				},
-				expect: `schemaVersion: scenario/v1
-title: echo
-vars:
-  message: null
 steps:
 - title: POST /say
   protocol: http
@@ -598,6 +560,36 @@ steps:
 				expect: fmt.Sprintf(`failed to load scenarios: ytt failed:
 - undefined: msg
     %s/testdata/ytt_invalid.yaml:4 |   message: #@ msg`, wd),
+			},
+			"disable ytt integration": {
+				config: &schema.Config{
+					Scenarios: []string{"testdata/ytt.yaml"},
+				},
+				expect: `failed to load scenarios: failed to decode YAML: [5:3] string was used where mapping is expected
+       2 | schemaVersion: scenario/v1
+       3 | title: echo
+       4 | vars:
+    >  5 |   message #@ msg: null
+             ^
+       6 | steps:
+       7 | - title: POST /say
+       8 |   protocol: http
+       9 |  `,
+			},
+			"invalid but disable ytt integration": {
+				config: &schema.Config{
+					Scenarios: []string{"testdata/ytt_invalid.yaml"},
+				},
+				expect: `failed to load scenarios: failed to decode YAML: [4:3] string was used where mapping is expected
+       1 | schemaVersion: scenario/v1
+       2 | title: echo
+       3 | vars:
+    >  4 |   message #@ msg: null
+             ^
+       5 | steps:
+       6 | - title: POST /say
+       7 |   protocol: http
+       8 |  `,
 			},
 		}
 		for name, test := range tests {

--- a/schema/load.go
+++ b/schema/load.go
@@ -1,11 +1,11 @@
 package schema
 
 import (
-	"bytes"
 	"fmt"
 	"io"
 	"os"
 	"path/filepath"
+	"strings"
 
 	"github.com/fatih/color"
 	"github.com/goccy/go-yaml"
@@ -180,19 +180,41 @@ func findAllfiles(paths ...string) ([]string, error) {
 }
 
 func loadScenariosFromFileAST(f *ast.File) ([]*Scenario, error) {
-	var buf bytes.Buffer
-	dec := yaml.NewDecoder(&buf, yaml.UseOrderedMap(), yaml.Strict())
+	// workaround to avoid the issue #304
+	r := strings.NewReader(f.String())
+	dec := yaml.NewDecoder(r, yaml.UseOrderedMap(), yaml.Strict())
 	var scenarios []*Scenario
-	for _, doc := range f.Docs {
+	var i int
+	for {
 		var s Scenario
-		if err := dec.DecodeFromNode(doc.Body, &s); err != nil {
+		if err := dec.Decode(&s); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
 			return nil, fmt.Errorf("failed to decode YAML: %w", err)
 		}
 		s.filepath = f.Name
-		s.Node = doc.Body
+		if i < len(f.Docs) {
+			s.Node = f.Docs[i].Body
+		}
 		scenarios = append(scenarios, &s)
+		i++
 	}
 	return scenarios, nil
+
+	// var buf bytes.Buffer
+	// dec := yaml.NewDecoder(&buf, yaml.UseOrderedMap(), yaml.Strict())
+	// var scenarios []*Scenario
+	// for _, doc := range f.Docs {
+	// 	var s Scenario
+	// 	if err := dec.DecodeFromNode(doc.Body, &s); err != nil {
+	// 		return nil, fmt.Errorf("failed to decode YAML: %w", err)
+	// 	}
+	// 	s.filepath = f.Name
+	// 	s.Node = doc.Body
+	// 	scenarios = append(scenarios, &s)
+	// }
+	// return scenarios, nil
 }
 
 type loadOption struct {

--- a/schema/load_test.go
+++ b/schema/load_test.go
@@ -17,6 +17,7 @@ import (
 	"github.com/zoncoen/scenarigo/assert"
 	"github.com/zoncoen/scenarigo/context"
 	"github.com/zoncoen/scenarigo/protocol"
+	"github.com/zoncoen/scenarigo/protocol/http"
 )
 
 type testProtocol struct {
@@ -564,6 +565,49 @@ func TestMarshalYAML(t *testing.T) {
 	}
 
 	if got, expect := buf.String(), string(b); got != expect {
+		dmp := diffmatchpatch.New()
+		diffs := dmp.DiffMain(expect, got, false)
+		t.Errorf("differs:\n%s", dmp.DiffPrettyText(diffs))
+	}
+}
+
+func Test_Issue304(t *testing.T) {
+	http.Register()
+	yml := `title: get scenarigo repository
+steps:
+- title: expect 404
+  protocol: http
+  request:
+    method: GET
+    url: https://api.github.com/repos/zoncoen/aaaaaaaaaaaaa
+    body:
+      # message: this comment does evil
+  expect:
+    code: Not Found
+`
+	scenarios, err := LoadScenariosFromReader(strings.NewReader(yml))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if l := len(scenarios); l != 1 {
+		t.Fatalf("unexpected length %d", l)
+	}
+	b, err := yaml.Marshal(scenarios[0])
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	expect := `title: get scenarigo repository
+steps:
+- title: expect 404
+  protocol: http
+  request:
+    method: GET
+    url: https://api.github.com/repos/zoncoen/aaaaaaaaaaaaa
+  expect:
+    code: Not Found
+`
+	if got := string(b); got != expect {
 		dmp := diffmatchpatch.New()
 		diffs := dmp.DiffMain(expect, got, false)
 		t.Errorf("differs:\n%s", dmp.DiffPrettyText(diffs))

